### PR TITLE
Store reference_cell_types in Triangulation

### PR DESCRIPTION
**Pull request by peterrum**

### CODE REVIEW

File: include/deal.II/grid/tria.h [hunk 1]
```diff
@@ -3359,6 +3359,18 @@ class Triangulation : public Subscriptor
...
+   * TODO
+   */
+  const std::vector<ReferenceCell::Type> &
+  get_reference_cell_types() const;
+
+  /**
+   * TODO
+   */
+  bool
+  all_reference_cell_types_are_hyper_cube() const;
```
Comment (peterrum) [inline include/deal.II/grid/tria.h:3372]: @jppelteret What do you think about this function? Is this what you had in mind at our last chat?

File: source/grid/tria.cc [hunk 2]
```diff
@@ -12771,6 +12771,39 @@ Triangulation<dim, spacedim>::update_periodic_face_map()
 }
 
 
+template <int dim, int spacedim>
```
Comment (kronbichler) [inline source/grid/tria.cc:12774]: Can you use 3 empty lines between the functions here and below?

Comment (peterrum) [reply source/grid/tria.cc:12774]: I can do that. But 2 lines are the norm in tria.cc, which is why I did not use 3 lines.

Comment (kronbichler) [reply source/grid/tria.cc:12774]: I saw both in the file with maybe a slight tendency to 2 lines, but 3 lines are also quite common; the reason why I mentioned it is that the place where you inserted the code used to have 3 lines of empty space :wink:

Comment (peterrum) [reply source/grid/tria.cc:12775]: Done!

**Review by kronbichler — APPROVED** [review]

**Comment by drwells** [comment]
I find using a vector here confusing - when I first read this I thought we could index `reference_cell_types` in a meaningful way. Is there an advantage to using a vector over just using a set?

File: include/deal.II/grid/tria.h (hunk 1 above)
Comment (jppelteret) [reply include/deal.II/grid/tria.h:3374]: Yes, precisely @peterrum. Thanks for adding it! I would want to use it for much the same purposes as you've done [here](https://github.com/dealii/dealii/pull/11140/files#diff-a3f9b6c7a51a53e5c8dc202a4dc8b8ab87f574023fb5f04dd3d2d87ecc62d68bR9814-R9818).

Comment (peterrum) [reply include/deal.II/grid/tria.h:3374]: Thanks for the feedback! It took some time until I had time to work in this issue ;)

**Comment by drwells** [comment]
@peterrum pointed out to me that we already store manifold and boundary ids in this way so this approach is best.

**Pull request merged**